### PR TITLE
[processing] port service area algorithms to C++

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Port of the service area Processing network analysis algorithms to C++. Follow up #7505.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+QStringList QgsServiceAreaFromLayerAlgorithm::tags() const
+{
+  return QObject::tr( "network,service,area,shortest,fastest" ).split( ',' );
+}
+
+QString QgsServiceAreaFromLayerAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm computes service areas from the set of start points defined by point layer using travel time or distance." );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: There was more detail in the original yaml file, why was this removed?

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+      {
+        // vertex is too expensive, discard
+        continue;
+      }
+
+      vertices.insert( j );
+      startPoint = graph->vertex( j ).point();
+
+      // find all edges coming from this vertex
+      for ( int edgeId : graph->vertex( j ).outgoingEdges() )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]:
```suggestion
      const QVector< int > outgoingEdges = graph->vertex( j ).outgoingEdges() ;
      for ( int edgeId : outgoingEdges )
```

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+          areaPoints.push_back( interpolatedEndPoint );
+          lines.push_back( QgsPolylineXY() << startPoint << interpolatedEndPoint );
+        }
+      } // edges
+    } // costs
+
+    // convert to list and sort to maintain same order of points between algorithm runs
+    QList< int > verticesList = vertices.toList();
+    std::sort( verticesList.begin(), verticesList.end() );
+    for ( int v : verticesList )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: areaPoints.reserve( verticesList.size());

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+      feat.setGeometry( geomPoints );
+      attributes = sourceAttributes.value( i + 1 );
+      attributes << QStringLiteral( "within" ) << origPoint;
+      feat.setAttributes( attributes );
+      pointsSink->addFeature( feat, QgsFeatureSink::FastInsert );
+
+      if ( includeBounds )
+      {
+        QgsMultiPointXY upperBoundary, lowerBoundary;
+        QVector< int > nodes;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: nodes.reserve(costs.size());

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+          {
+            vertexId = graph->edge( tree.at( v ) ).fromVertex();
+            if ( costs.at( vertexId ) <= travelCost )
+            {
+              nodes.push_back( v );
+            }
+          }
+        } // costs
+
+        for ( int n : nodes )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]:
```suggestion
        for ( int n : qgis::as_const(nodes) )
```

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        {
+          upperBoundary.push_back( graph->vertex( graph->edge( tree.at( n ) ).toVertex() ).point() );
+          lowerBoundary.push_back( graph->vertex( graph->edge( tree.at( n ) ).fromVertex() ).point() );
+        } // nodes
+
+        QgsGeometry geomUpper = QgsGeometry::fromMultiPointXY( upperBoundary );
+        QgsGeometry geomLower = QgsGeometry::fromMultiPointXY( lowerBoundary );
+
+        feat.setGeometry( geomUpper );
+        attributes = sourceAttributes.value( i + 1 );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: The python version didn't have +1 here -- is that intended?

File: src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+        points.push_back( interpolatedEndPoint );
+        lines.push_back( QgsPolylineXY() << edgeStart << interpolatedEndPoint );
+      }
+    } // edges
+  } // costs
+
+  // convert to list and sort to maintain same order of points between algorithm runs
+  QList< int > verticesList = vertices.toList();
+  std::sort( verticesList.begin(), verticesList.end() );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp]: reserve points size here too

**Review by nyalldawson — APPROVED** [review]
Looking great!

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp (hunk 1 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: Seems they were added *after* initial port. Anyway I copied them now.

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp (hunk 2 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: Hmm, maybe I'm wrong but we can not assign QList to QVector, can we? Replaced with `const QList<int>`, hope this is ok too.

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp (hunk 3 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: Done

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp (hunk 4 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: Done

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp (hunk 5 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: Done

File: src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp (hunk 7 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafrompoint.cpp]: Done

File: src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp (hunk 6 above)
Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp]: My bad. Fixed.

File: src/analysis/processing/qgsalgorithmserviceareafrompoint.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/***************************************************************************
+                         qgsalgorithmserviceareafrompoint.h
+                         ---------------------
+    begin                : JUly 2018
```
Comment (roya0045) [inline src/analysis/processing/qgsalgorithmserviceareafrompoint.h]: 2018?
